backend/src/apiserver/server/pipeline_upload_server_test.go

```diff
@@ -553,3 +447,43 @@ func TestDefaultUpdatedPipelineVersion(t *testing.T) {
 	assert.Nil(t, err)
 	assert.Equal(t, pipeline.DefaultVersionId, fakeVersionUUID)
 }
+
+func uploadeFromFile() (*resource.FakeClientManager, *resource.ResourceManager, PipelineUploadServer, *bytes.Buffer, *multipart.Writer, io.Writer) {
+	clientManager, resourceManager, server, bytesBuffer, writer := setupBase()
+	part, _ := writer.CreateFormFile("uploadfile", "hello-world.yaml")
+	io.Copy(part, bytes.NewBufferString("apiVersion: argoproj.io/v1alpha1\nkind: Workflow"))
+	writer.Close()
+	return clientManager, resourceManager, server, bytesBuffer, writer, part
+}
+
+func uploadPipeline(method string, url string) (*resource.FakeClientManager, *resource.ResourceManager, *http.Request, *httptest.ResponseRecorder, PipelineUploadServer, *bytes.Buffer, *multipart.Writer, http.HandlerFunc) {
```

Comment:
This is already improving existing tests a lot, so it's fine if you want to merge like this too.

Anyway, I'm giving a general comment if you want to improve it further:
It's better to write a utility function that does one thing well and can really encapsulate the action, because that makes it easy to **compose** multiple function calls to test any flexible things.
Right now, the `uploadPipeline` function also sets up the test, so we cannot call this method twice in one test case and we still have places that require a lot of boilerplate code for uploading the pipeline the second time.
Also, it returns so many variables making it really hard to understand what it is doing under the hood.

So I'd suggest creating a few `setup` methods that just do the setup and then the `uploadPipeline` method should take required inputs from the setup method, so tests can setup once and upload multiple times with different parameters.

Reply:
I didn't look through the complex set up, so I guess you got your difficulties achieving this, feel free to discuss that with me.

Reply:
Thanks for the feedback, I will break it up and fix it :)